src/api/utils/index.js

```diff
@@ -1,31 +1,117 @@
+import 'babel-polyfill';
 import imm from 'object-path-immutable';
+import Registry from '../../registry';
 
 export apiClient from './api-client';
 
+/**
+ * @module
+ * @private
+ */
+
+/**
+ * @example
+ * incrementProperty(state, 'isReading')
+ * @param {Object} state Object on which the key is incremented
+ * @param {string} key Key to increment
+ */
 export const incrementProperty = (state, key) => (
   imm.set(state, key, state[key] + 1)
 );
 
+/**
+ * @example
+ * decrementProperty(state, 'isReading')
+ * @param {Object} state Object on which the key is decremented
+ * @param {string} key Key to decrement
+ */
 export const decrementProperty = (state, key) => (
   imm.set(state, key, state[key] - 1)
 );
 
-export const resourcesArray = (data) => (
+/**
+ * @param {Array|Object} data
+ */
+export const resourcesArray = (data = []) => (
   Array.isArray(data) ? data : [data]
 );
 
+/**
+ * @example
+ * const obj = { foo: bar: { baz: 'hi' } }
+ * hasProperty(obj, 'foo', 'bar', 'baz') // => 'hi'
+ * hasProperty(obj, 'a', 'b', 'c') // => undefined
+ *
+ * @param {Object} object Object to traverse
+ * @param {String[]} keys Keys to look up
+ */
+export const hasProperty = (object, ...keys) => (
+  keys.reduce((acc, key) => {
+    if (acc === undefined) return acc;
+    return acc[key];
+  }, object)
+);
+
+/**
+ * Creates reverse relationships for data merged into the state
+ * @param {Object} state
+ * @param {Object} data JSON-API data object
+ * @param {Object} data.id
+ * @param {Object} data.type
+ * @param {Object} data.relationships
+ */
+export const createReverseRelationships = (state, newState, { id, type, relationships = {} }) => {
+  Object.values(relationships).forEach(({ data }) => {
+    const children = resourcesArray(data);
+    const childModel = Registry.get(children[0].type);
+    const childData = { type, id };
+    const { key, isArray } = childModel.relationships[type];
+
+    children.forEach(({ type: relType, id: relId }) => {
+      const objectPath = ['resources', relType, relId, 'relationships', key];
+      const existingRelationships = resourcesArray(hasProperty(state, ...objectPath));
+
+      if (!isArray) {
+        newState.set([...objectPath, 'data'], childData);
```

Comment:
There are two ways of using the immutable object:
1. `imm(object)` - all subsequent calls mutate that object, `#value()` gets the new object back
2. `imm.set(object, path, data)` - returns new object